proto_proxy.c

```diff
@@ -0,0 +1,2310 @@
+    }
+    return b+1; // FIXME: do the -1 just for ketama and remove from internal code?
+}
+
+// stack = [pool, option]
+static int mcplib_jump_hash_new(lua_State *L) {
+    uint64_t seed = 0;
+    const char *seedstr = NULL;
+    size_t seedlen = 0;
+
+    luaL_checktype(L, 1, LUA_TTABLE);
+    lua_Unsigned buckets = lua_rawlen(L, 1);
+
+    int argc = lua_gettop(L);
+    if (argc > 1) {
+        // options supplied. to be specified as a table.
+        // { seed = "foo" }
+        luaL_checktype(L, 2, LUA_TTABLE);
+        // FIXME: adjust so we ensure/error on this being a string?
+        if (lua_getfield(L, 2, "seed") != LUA_TNIL) {
+            seedstr = lua_tolstring(L, -1, &seedlen);
+            seed = XXH3_64bits(seedstr, seedlen);
+        } else {
+            dump_stack(L);
+        }
+        lua_pop(L, 1);
+    }
+
+    mcplib_jump_hash_t *jh = lua_newuserdatauv(L, sizeof(mcplib_jump_hash_t), 0);
+    // TODO: check jh.
+
+    // don't need to loop through the table at all, just need its length.
+    // could optimize startup time by adding hints to the module for how to
+    // format pool (ie; just a total count or the full table)
+    jh->seed = seed;
+    jh->buckets = buckets;
+    jh->phc.ctx = jh;
+    jh->phc.selector_func = mcplib_jump_hash_get_server;
+
+    lua_pushlightuserdata(L, &jh->phc);
+
+    // - return [UD, lightuserdata]
+    return 2;
+}
+
+static int mcplib_open_jump_hash(lua_State *L) {
+    const struct luaL_Reg jump_f[] = {
+        {"new", mcplib_jump_hash_new},
+        {NULL, NULL},
+    };
+
+    luaL_newlib(L, jump_f);
+
+    return 1;
+}
+
+/*** END jump consistent hash library ***/
+
+// Creates and returns the top level "mcp" module
+int proxy_register_libs(LIBEVENT_THREAD *t, void *ctx) {
+    lua_State *L = ctx;
+    // TODO: stash into a table with weak references?
+    // then if no pools/code has references still, can ditch?
+    // TODO: __gc
+    const struct luaL_Reg mcplib_backend_m[] = {
+        {"set", NULL},
+        {NULL, NULL}
+    };
+
+    const struct luaL_Reg mcplib_request_m[] = {
+        {"command", mcplib_request_command},
+        {"key", mcplib_request_key},
+        {"__tostring", NULL},
+        {"__gc", mcplib_request_gc},
+        {NULL, NULL}
+    };
+
+    const struct luaL_Reg mcplib_response_m[] = {
+        {"ok", mcplib_response_ok},
+        {"__gc", mcplib_response_gc},
+        {NULL, NULL}
+    };
+
+    // TODO: __gc
+    const struct luaL_Reg mcplib_hash_selector_m[] = {
+        {"__call", mcplib_hash_selector_call},
+        {NULL, NULL}
+    };
+
+    const struct luaL_Reg mcplib_f [] = {
+        {"hash_selector", mcplib_hash_selector},
+        {"backend", mcplib_backend},
+        {"request", mcplib_request},
+        {"attach", mcplib_attach},
+        {NULL, NULL}
+    };
+
+    // TODO: function + loop.
+    luaL_newmetatable(L, "mcp.backend");
+    lua_pushvalue(L, -1); // duplicate metatable.
```

Comment:
why are we pushing the metatable twice? 

Isn't the stack looking like   **bottom** [ <mcp.backend>, <mcp.backend> ] **top** after this line?
And after registering the functions to this metatable, you are calling lua_pop(L,1), which will pop only 1 element leaving stack as follows: **bottom** [ <mcp.backend>] **top** ?

Is my understanding correct?

After line 2283, the stack looks like the following - 

**bottom** [ <mcp.backend>, <mcp.request>, <mcp.response>, <mcp.hash_selector>] **top**

Reply:
Did you change something? After 2283 the stack is empty.

Reply:
```
      luaL_newmetatable(L, "mcp.backend"); <-- [table]
      lua_pushvalue(L, -1); // duplicate metatable. <-- [table, table] (second is a ref to first)
      lua_setfield(L, -2, "__index"); // mt.__index = mt <-- [table] (popped after setting into table)
      luaL_setfuncs(L, mcplib_backend_m, 0); // register methods <-- [table] (functions assigned, returns in orig state)
      lua_pop(L, 1); <-- [] (table removed, moves onto next metatable)
```

Reply:
Got it, didn't realize lua_setfield does the pop operation. So the duplicate insertion is to index it essentially.
It is clear now.